Sources/Basic/PathShims.swift

```diff
@@ -155,86 +151,52 @@ public func walk(_ path: AbsolutePath, recursively: Bool = true) -> RecursibleDi
  are silently ignored. So keep that in mind.
  
- - Warning: If path doesn’t exist or cannot be entered this generator will
- be empty. It is up to you to check `path` is valid before using this
- function.
- 
  - Warning: Symbolic links that point to directories are *not* followed.
  
  - Note: returning `false` from `recursing` still produces that directory
  from the generator; just not its contents.
  */
-public func walk(_ path: AbsolutePath, recursing: (AbsolutePath) -> Bool) -> RecursibleDirectoryContentsGenerator {
-    return RecursibleDirectoryContentsGenerator(path: path, recursionFilter: recursing)
-}
-
-/**
- A generator for a single directory’s contents
- */
-private class DirectoryContentsGenerator: IteratorProtocol {
-    private let dirptr: DirHandle?
-    fileprivate let path: AbsolutePath
-    
-    fileprivate init(path: AbsolutePath) {
-        dirptr = libc.opendir(path.asString)
-        self.path = path
-    }
-    
-    deinit {
-        if let openeddir = dirptr { closedir(openeddir) }
-    }
-    
-    func next() -> dirent? {
-        guard let validdir = dirptr else { return nil }  // yuck, silently ignoring the error to maintain this pattern
-        
-        while true {
-            var entry = dirent()
-            var result: UnsafeMutablePointer<dirent>? = nil
-            guard readdir_r(validdir, &entry, &result) == 0 else { continue }
-            guard result != nil else { return nil }
-            
-            switch (entry.d_name.0, entry.d_name.1, entry.d_name.2) {
-            case (46, 0, _):   // "."
-                continue
-            case (46, 46, 0):  // ".."
-                continue
-            default:
-                return entry
-            }
-        }
-    }
+public func walk(_ path: AbsolutePath, fileSystem: FileSystem = localFileSystem, recursing: (AbsolutePath) -> Bool) throws -> RecursibleDirectoryContentsGenerator {
+    return try RecursibleDirectoryContentsGenerator(path: path, fileSystem: fileSystem, recursionFilter: recursing)
 }
 
 /**
  Produced by `walk`.
  */
 public class RecursibleDirectoryContentsGenerator: IteratorProtocol, Sequence {
-    private var current: DirectoryContentsGenerator
+    private var current: (path: AbsolutePath, iterator: IndexingIterator<[String]>)
     private var towalk = [AbsolutePath]()
+
     private let shouldRecurse: (AbsolutePath) -> Bool
+    private let fileSystem: FileSystem
     
-    private init(path: AbsolutePath, recursionFilter: (AbsolutePath) -> Bool) {
-        current = DirectoryContentsGenerator(path: path)
+    private init(path: AbsolutePath, fileSystem: FileSystem, recursionFilter: (AbsolutePath) -> Bool) throws {
+        self.fileSystem = fileSystem
+        // FIXME: getDirectoryContents should have an iterator version.
+        current = (path, try fileSystem.getDirectoryContents(path).makeIterator())
         shouldRecurse = recursionFilter
     }
-    
+
     public func next() -> AbsolutePath? {
```

Comment:
I find this function rather hard to read (and trust), but it isn't related to the content of this patch.